### DISCUSSION TITLE
Styling Shell with C# (No XAML) -- How to?

### DISCUSSION THREAD

**Discussion by devonuto** [CommunityToolkit/Maui.Markup#61] [Q&A] [answered]
Hi everyone, started building a MAUI app in straight C#, and running into styling issues.  Specifically using Microsoft.Maui.Controls.Shell, I'm having trouble getting the styling of the flyoutitem labels, and tabs on the shell.  They also appear in different colours depending on whether the Maui app is run on Android, or in Windows.  I can seem to make sense of it.

I've been trying to find how to write my own default style and apply it to my own extended flyoutitem class for the labels/background etc, but can't see how to apply setters to the style.  I also need to have visual state setters etc.

Has anyone got any demo apps in github that do this?

**Comment by bijington** [comment, marked as answer]
We don't have any styling set around Shell in a non-XAML application but we do have a sample non-XAML application. Take a look at our Markup sample and let me know if that helps:

https://github.com/CommunityToolkit/Maui.Markup/tree/main/samples

**Reply by TheCodeTraveler** [reply]
Adding Shell support sounds like a great idea for a New Feature Proposal for CommunityToolkit.Maui.Markup!

**Reply by devonuto** [reply]
I've got the styling mostly working, but still can't seem to get font images on the tabs to properly align in the vertical center of the tab (no textalignment seems to work).

```cs
public AppShell()
{
    Style = new Style(typeof(Element))
    {
        ApplyToDerivedTypes = true,
        Setters =
        {
            new Setter { Property = Shell.BackgroundColorProperty, Value = ColourConstants.BackgroundColour },
            new Setter { Property = Shell.DisabledColorProperty, Value = ColourConstants.DesaturatedDark },
            new Setter { Property = Shell.FlyoutBackgroundColorProperty, Value = ColourConstants.Desaturated },
            new Setter { Property = Shell.ForegroundColorProperty, Value = ColourConstants.FontColour },
            new Setter { Property = Shell.TabBarBackgroundColorProperty, Value = ColourConstants.Primary },
            new Setter { Property = Shell.TabBarDisabledColorProperty, Value = ColourConstants.DesaturatedDark },
            new Setter { Property = Shell.TabBarForegroundColorProperty, Value = ColourConstants.ButtonText },
            new Setter { Property = Shell.TabBarTitleColorProperty, Value = ColourConstants.ButtonText },
            new Setter { Property = Shell.TabBarUnselectedColorProperty, Value = ColourConstants.Unselected },
            new Setter { Property = Shell.TitleColorProperty, Value = ColourConstants.FontColour },
            new Setter { Property = Shell.UnselectedColorProperty, Value = ColourConstants.DesaturatedDark },
            new Setter { Property = Shell.FlyoutWidthProperty, Value = 200 },
        }
    }
}
```

**Reply by TheCodeTraveler** [reply]
@devonuto - for writing styles in C#, I _highly_ recommend using the [`CommunityToolkit.Maui.Markup` NuGet Package](https://www.nuget.org/packages/CommunityToolkit.Maui.Markup/). 

Here's the docs its `Style<T>` implementation: [https://docs.microsoft.com/dotnet/communitytoolkit/maui/markup/extensions/style](https://docs.microsoft.com/dotnet/communitytoolkit/maui/markup/extensions/style?WT.mc_id=dotnet-0000-bramin)

It'll allow you to significantly reduce the complexity of your code to this:

```cs
public AppShell()
{
    Style = new Style<Element>()
                     .ApplyToDerivedTypes(true)
                     .Add(Shell.BackgroundColorProperty, ColourConstants.BackgroundColour)
                     .Add(Shell.DisabledColorProperty, ColourConstants.DesaturatedDark)
                     .Add(Shell.FlyoutBackgroundColorProperty, ColourConstants.Desaturated)
                     .Add(Shell.ForegroundColorProperty, ColourConstants.FontColour)
                     .Add(Shell.TabBarBackgroundColorProperty, ColourConstants.Primary)
                     .Add(Shell.TabBarDisabledColorProperty, ColourConstants.DesaturatedDark)
                     .Add(Shell.TabBarForegroundColorProperty, ColourConstants.ButtonText)
                     .Add(Shell.TabBarTitleColorProperty, ColourConstants.ButtonText)
                     .Add(Shell.TabBarUnselectedColorProperty, ColourConstants.Unselected)
                     .Add(Shell.TitleColorProperty, ColourConstants.FontColour)
                     .Add(Shell.UnselectedColorProperty, ColourConstants.DesaturatedDark)
                     .Add(Shell.FlyoutWidthProperty, 200);
}
```

**Reply by devonuto** [reply]
@brminnick I was essentially just trying to copy how the XAML was structured to get it to work.  Will have a read of that, thanks.

**Reply by devonuto** [reply]
PS: when using the < > add code, for me it just adds lines of comments... couldn't get it to format nicely like that.

**Reply by TheCodeTraveler** [reply]
GitHub uses Markdown to format text.

Here's how to format a Code Block in Markdown: https://www.markdownguide.org/extended-syntax/#fenced-code-blocks

**Reply by TheCodeTraveler** [reply]
You need 3 back ticks to create a [fenced code block](https://www.markdownguide.org/extended-syntax/#fenced-code-blocks).

I've edited it for you to demonstrate  how it works.

**Reply by devonuto** [reply]
One question now, is how do you do "add" for device specific?

I was previously doing this:

    if (DeviceInfo.Platform == DevicePlatform.WinUI)
    {
        CollectionViewSyle.Setters.Add(new Setter { Property = VisualElement.MaximumWidthRequestProperty, Value = 800 });
        GridStyle.Setters.Add(new Setter { Property = VisualElement.MaximumWidthRequestProperty, Value = 800 });
        RefreshViewStyle.Setters.Add(new Setter { Property = VisualElement.MaximumWidthRequestProperty, Value = 800 });
        LabelStyle.Setters.Add(new Setter { Property = Label.FontSizeProperty, Value = 22 });
        ConverterLabelStyle.Setters.Add(new Setter { Property = Label.FontSizeProperty, Value = 22 });
    }

I don't want to specify those for mobile devices, so is it okay to keep adding additional setters this way? I assume the end result is the same?

**Reply by TheCodeTraveler** [reply]
Yup, that's fine.

To leverage `CommuntyToolkit.Maui.Markup`, you'd use the same `Add` method:

```cs
if (DeviceInfo.Platform == DevicePlatform.WinUI)
{
    CollectionViewSyle.Add(VisualElement.MaximumWidthRequestProperty, 800);
    GridStyle.Add(VisualElement.MaximumWidthRequestProperty, 800);
    RefreshViewStyle.Add(VisualElement.MaximumWidthRequestProperty, 800);
    LabelStyle.Add(Label.FontSizeProperty, 22);
    ConverterLabelStyle.Add(Label.FontSizeProperty, 22);
}
```

**Reply by devonuto** [reply]
@brminnick That doesn't seem to work. e.g.:

    public static readonly Style CollectionViewSyle = new Style<CollectionView>()
        .ApplyToDerivedTypes(true)
        .Add(View.HorizontalOptionsProperty, LayoutOptions.Center)
        .Add(SelectableItemsView.SelectionModeProperty, SelectionMode.Single)
        .Add(View.VerticalOptionsProperty, LayoutOptions.Center);
        
         if (DeviceInfo.Platform == DevicePlatform.WinUI)
        {
            CollectionViewSyle.Add(VisualElement.MaximumWidthRequestProperty, 800 );
        }
        
Error CS1929 'Style' does not contain a definition for 'Add' and the best extension method overload 'GridExtensions.Add(Grid, IView, int, int)' requires a receiver of type 'Grid'

**Reply by devonuto** [reply]
Just noticed changing that to this fixes it:

    public static readonly Style<CollectionView> CollectionViewSyle = new Style<CollectionView>()
        .ApplyToDerivedTypes(true)
        .Add(View.HorizontalOptionsProperty, LayoutOptions.Center)
        .Add(SelectableItemsView.SelectionModeProperty, SelectionMode.Single)
        .Add(View.VerticalOptionsProperty, LayoutOptions.Center);

**Reply by devonuto** [reply]
@brminnick  is there a CommuntyToolkit.Maui.Markup way of adding visual state setters in C#?

**Reply by devonuto** [reply]
Also can't figure out how to add a class for overriding flyout styles in this method?

<Style Class="FlyoutItemLayoutStyle" TargetType="Layout" ApplyToDerivedTypes="True">

**Reply by devonuto** [reply]
This is the only way I have found that works for adding setters so far (by defining my own Grid type).

    public class MyGrid : Grid
    {
        public MyGrid()
        {
            VisualStateManager.SetVisualStateGroups(
                this,
                new VisualStateGroupList()
                {
                    new VisualStateGroup()
                    {
                        States =
                        {
                            new VisualState()
                            {
                                Name = "Selected",
                                Setters =
                                {
                                    new Setter { Property = Grid.BackgroundColorProperty, Value = ColourConstants.Primary }
                                }
                            }
                        }
                    }
                }
            );
        }
    }

**Reply by pictos** [reply]
@devonuto can you show the code that you would like to write in order to define the VisualStateManager?

**Reply by devonuto** [reply]
@pictos Well if you have this for adding setters:

```
public static readonly Style<Grid> GridSyle = new Style<Grid>()
    .ApplyToDerivedTypes(true)
    .Add(Grid.BackgroundColorProperty, ColourConstants.BackgroundColour);
```

Extending it to add visual states, if possible maybe something simple like this:

```
public static readonly Style<Grid> GridSyle = new Style<Grid>()
    .ApplyToDerivedTypes(true)
    .Add(Grid.BackgroundColorProperty, ColourConstants.BackgroundColour)
    .Add(VisualStateManager.VisualStateGroups, new VisualState(VisualStateType.Selected)
    {
        Setters.Add(Grid.BackgroundColorProperty, ColourConstants.Primary)
    });
```

But I don't know if that would be too simplified. i.e do we need to define groups first, or just assume all setters are added to the same group etc

**Reply by pictos** [reply]
@devonuto do you mind opening a discussion asking for that extension and using the API that you showed here as a reference?

You just did an API design for us, thanks <3

**Comment by TheCodeTraveler** [comment]
Just FYI - since this is focused on C# + MAUI, I transferred this Discussion to the CommunityToolkit/Maui.Markup repo 👍

**Comment by CliffAgius** [comment]
Closed and answered
